### ISSUE TITLE
Time parsing tests fail when localtime is in Daylight Savings

### ISSUE THREAD

**Issue by orner** [akrennmair/newsbeuter#369]
The seven asserts in the unit tests that depend on time parsing fail by being off by one hour, when localtime is in daylight savings time.  These test will pass when local time is set to January, when daylight savings in not in effect, or location set to Phoenix, which doesn't have a daylight savings.  I would suppose using UTC would also work.

A quick check of the code seems to be around rss/rss_parser.cpp:118, where some variables reference PST (Pacific Standard) and others have PDT (Pacific Daylight).  I'll have to look more closely to determine what assumptions the different functions are making.

[datetestresults.txt](https://github.com/akrennmair/newsbeuter/files/504542/datetestresults.txt)

**Comment by Minoru** [comment]
Thanks for filing this, but I was hoping for some specific steps to reproduce that :)

**Comment by orner** [comment]
The attached file is my full output when I run test/test, since I'm on the US West Coast.  You can reproduce this by setting your local timezone to "Los Angeles", then running the tests.  That will change your timezone to "PST", however, since we're in Daylight Savings time, some of the functions will return "PDT".  I would imagine this can be reproduced by setting your local timezone to anyplace that has daylight savings.

**Comment by Minoru** [comment]
Yeah, I got the point. You didn't do anything wrong, it's just that you could do even better if you listed specific commands that let anyone easily replicate the bug. Like this:

```
$ make PROFILE=1 test

# The following passes without error
$ (cd test; TZ=US/Arizona ./test)

# The following fails with the output you showed above
$ (cd test; TZ=US/Pacific ./test)
```

Anyway, back to the bug.

[The spec](https://www.w3.org/TR/1998/NOTE-datetime-19980827) doesn't explicitly mention DST, but I'd expect the timezone offset accounts for it already: "A time zone offset of "+hh:mm" indicates that the date/time uses a local time zone which is "hh" hours and "mm" minutes ahead of UTC".

The culprit is indeed in `rss/rss_parser.cpp:118`. I'm looking into ways to properly fix it, but if you already know the answer, please share!

**Comment by orner** [comment]
Hmm, that would be a better way to put it.  I was screwing around with the timezone file for my adjustments.  :)

Thanks for the help.  This is new environment for me, so I'm still working my way through it.  I'll keep this in mind.  There are always new techniques to learn!

**Comment by Minoru** [comment]
I give up for now. This issue is free for the taking.

**Comment by orner** [comment]
Ugh.  This took some working in gdb to really watch what's going on.

The function is taking a string, converting it into a tm structure, converting that to the number of seconds (`time_t`) using `mktime()`, adding the offset, then converting back to tm structure to be printed in the correct format.  Without a better date/time library, `mktime`/`localtime`/`gmtime` is all that's really available.

`mktime()` is the only real function to convert from the tm structure with explicit year/month/day to a number of seconds, so the offset can be added.  However, [these docs](http://en.cppreference.com/w/cpp/chrono/c/mktime) point out in the notes, that `tm->tm_isdst` needs to be set before being called, which it wasn't.  I set it to -1 so that it takes into account DST.  The inverse function is `localtime()`, which always takes into account DST.  If we're trying to convert the string to seconds, add and offset, then reconvert to string without a local timezone affecting the result, `localtime()` needs to be used in tandem with `mktime()`, and, they need to convert it the same way.

So, my fix sets the DST flag, converts with `mktime()`, adds the offset, then converts back with `localtime()`, and manually zeros out any offset that `localtime()` implies, and prints that.

Test cases have been added in `test/rss.cpp` to set TZ to be in different timezones and then do a time conversion.  Everything passes.
